flang/lib/Optimizer/Transforms/AddAliasTags.cpp

```diff
@@ -154,18 +154,17 @@ void AddAliasTagsPass::runOnAliasInterface(fir::FirAliasTagOpInterface op,
 
     // TBAA for global variables
   } else if (enableGlobals &&
-             source.kind == fir::AliasAnalysis::SourceKind::Global) {
-    mlir::SymbolRefAttr glbl = source.u.get<mlir::SymbolRefAttr>();
+             source.kind == fir::AliasAnalysis::SourceKind::Global && !source.isBoxData()) {
+    mlir::SymbolRefAttr glbl = source.origin.u.get<mlir::SymbolRefAttr>();
     const char *name = glbl.getRootReference().data();
     LLVM_DEBUG(llvm::dbgs().indent(2) << "Found reference to global " << name
                                       << " at " << *op << "\n");
     tag = state.getFuncTree(func).globalDataTree.getTag(name);
 
     // TBAA for SourceKind::Direct
-  } else if (enableDirect &&
-             source.kind == fir::AliasAnalysis::SourceKind::Direct) {
-    if (source.u.is<mlir::SymbolRefAttr>()) {
-      mlir::SymbolRefAttr glbl = source.u.get<mlir::SymbolRefAttr>();
+  } else if (enableDirect && source.isBoxData()) {
```

Comment:
Is it possible to pass this check when the source is `Indirect` or `Allocate`?  I think that would be incorrect.  Can we remove this `else if` for now?  @tblah, do you agree?

Reply:
Yes, this is not equivalent to what it was before.
This would be:
```
enableDirect && source.kind == SourceKind::Global && source.isBoxData()
```

Of course, if we remove SourceKind::Direct maybe the nomenclature for enableDirect could change as well. 
But the behavior should be equivalent.

Reply:
Thanks for spotting this Slava, yes you were right here and the update was needed.

My intention behind this `enableDirect` path was for certain spec benchmarks with a lot of boxed arrays as globals. I agree that some better naming to make this clearer.